Base/Python/slicer/util.py

```diff
@@ -2395,6 +2395,116 @@ def createProgressDialog(parent=None, value=0, maximum=100, labelText="", window
       setattr(progressIndicator, key, value)
   return progressIndicator
 
+def getPythonConsoleWidget():
+  import slicer
+  return slicer.util.mainWindow().pythonConsole().parent()
+
+from contextlib import contextmanager
+
+@contextmanager
+def peakPythonConsole(show=True):
+  """Show the Python console while the code in the context manager is being run.
+
+  The console stays visible only if it was visible already.
+
+  :param show: If show is False, the context manager has no effect.
+
+  .. code-block:: python
+
+    with slicer.util.peakPythonConsole():
+      slicer.util.pip_install('nibabel')
+
+  """
+  import slicer
+  if show:
+    console = getPythonConsoleWidget()
+    pythonVisible = console.visible
+    console.setVisible(True)
+    slicer.app.processEvents()
+  try:
+    yield
+  finally:
+    if show:
+      console.setVisible(pythonVisible)
+
+@contextmanager
+def showWaitCursor(show=True):
+  """Use a waiting cursor while the code in the context manager is being run.
+
+  :param show: If show is False, the context manager has no effect.
+
+  .. code-block:: python
+
+    import time
+
+    with slicer.util.showWaitCursor():
+      time.sleep(2)
+  """
+  import qt, slicer
+  if show:
+    qt.QApplication.setOverrideCursor(qt.Qt.WaitCursor)
+    slicer.app.processEvents()
+  try:
+    yield
+  finally:
+    if show:
+      qt.QApplication.restoreOverrideCursor()
+
+@contextmanager
+def messageContextManager(message, show=True):
```

Comment:
Some context manager have the suffix `ContextManager` and some other ones to not. We should be consistent.

Also, we already have one context manager implemented as a class and it is called `NodeModify` that can only be used as context manager and not as a callable function.

It looks like some of the proposed context manager should also be only context manager, hence these suggestions:

* `showWaitCursor`
  * rename to `WaitCursor` and convert into a class like `NodeModify`
 
* `tryWithErrorDisplay`
  * instead introduce a manager called `slicer.util.raises` with the possibility to choose with option to catch.
  * always hide the dialog with testing mode is enabled  (`slicer.app.testingEnabled()`)
  * inspired from [pytest](https://docs.pytest.org/en/6.2.x/assert.html#assertions-about-expected-exceptions)
 
* `messageContextManager`
  * renamed to `MessageDialog`
  * consider approaching this as a decorator ? 
  * integrate with testing mode  (`slicer.app.testingEnabled()`)

Implementation of `NodeModify`:

https://github.com/Slicer/Slicer/blob/aaa9f918a4ee4f35bf09e05c5d7e53d083b2e4e3/Base/Python/slicer/util.py#L1343-L1351

Reply:
Thanks, @jcfr. I hadn't seen `NodeModify`. What is the rationale for choosing decorator vs class?

> with the possibility to choose with option to catch.

I'm not sure what you mean here. Can you maybe sketch some pseudo code?

> integrate with testing mode

What do you mean by "integrate" here? I'm not sure what you're asking me to do.

Reply:
re: integrate with testing code

If tests are running on the CI, no dialog should be shown. Considering that when we run tests in CI, we pass the parameter `--testing`, the function `slicer.app.testingEnabled()` returns True .. and we should test for this to make sure no dialog are shown. Failing to do would require user input ... which is not possible when running tests.

Reply:
Oh, sure. The same as above, then. Thanks.

Reply:
I've turned those two generators into classes and integrated all of these with `testingEnabled`.

> instead introduce a manager called slicer.util.raises with the possibility to choose with option to catch.

I'm still not sure how this is related to pytest's raises, nor why we are only converting some of these into classes.

Reply:
>  I'm still not sure how this is related to pytest's raises

Instead of hard coding the general python class `Exception`, it was suggestion it is passed as a parameter.

Also , the function would be renamed from `tryWithErrorDisplay` to `raises`

and would be used like this:

```
with slicer.util.raises(Exception):
  [...]
```

Reply:
Generally speaking it also bad practice to catch the general exception, it should be specific one ...

Reply:
The goal of this manager is "try to do something and if you get _any_ error, show it to me". That's why I'm "catching" (not really, it's just to show the message on the dialog) the general exception. The idea is that the user of this context manager doesn't know a priori which exceptions may be raised.

Reply:
This is a relevant reading: https://stackoverflow.com/questions/4990718/about-catching-any-exception#4992124

If the goal is to catch the exception to let the user know about .. then we could consider re-raising it afterward to make sure it can be handled later by the application  ?

Reply:
The guy in the post is doing exactly what I've done...

I agree with the re-raising. I'll add it now.